### PR TITLE
AAE-26192 New task search API

### PR DESCRIPTION
**Pull request by DudaRobert**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [x] Tests for the changes have been added (for bug fixes / features)
> - [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [ ] Bugfix
> - [x] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)
https://hyland.atlassian.net/browse/AAE-26192

**What is the new behaviour?**
https://hyland.atlassian.net/browse/AAE-26192

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: lib/process-services-cloud/src/lib/process/process-list/services/process-task-list-cloud.service.ts [hunk 1]
```diff
@@ -19,13 +19,12 @@ import { Injectable } from '@angular/core';
 import { Observable, throwError } from 'rxjs';
 import { BaseCloudService } from '../../../services/base-cloud.service';
 import { map } from 'rxjs/operators';
-import { TaskListCloudServiceInterface } from '../../../services/task-list-cloud.service.interface';
 import { TaskQueryCloudRequestModel } from '../../../models/filter-cloud-model';
 import { TaskCloudNodePaging } from '../../../models/task-cloud.model';
 import { TaskListCloudSortingModel } from '../../../models/task-list-sorting.model';
 
 @Injectable({ providedIn: 'root' })
-export class ProcessTaskListCloudService extends BaseCloudService implements TaskListCloudServiceInterface {
```
Comment (DudaRobert) [inline lib/process-services-cloud/src/lib/process/process-list/services/process-task-list-cloud.service.ts:28]: `ProcessTaskListCloudService ` shouldn't implement that interface - it's not event shared between tasks and processes.

File: lib/process-services-cloud/src/lib/task/task-filters/services/task-filter-cloud.service.ts [hunk 2]
```diff
@@ -379,4 +380,29 @@ export class TaskFilterCloudService extends BaseCloudService {
     refreshFilter(filterKey: string): void {
         this.filterKeyToBeRefreshedSource.next(filterKey);
     }
+
+    private handleBackwardCompatibility(appName: string, key: string, filters: TaskFilterCloudModel[]): Observable<TaskFilterCloudModel[]> {
```
Comment (DudaRobert) [inline lib/process-services-cloud/src/lib/task/task-filters/services/task-filter-cloud.service.ts:384]: This method will filre at the start of the app, when we retrieve the filters from local storage or when we first populate them.
If we don't find the new keys in the retrieved filters, but we find their old counterparts we populate the new keys and leave the counterparts as they are.

File: lib/process-services-cloud/src/lib/task/task-filters/services/task-filter-cloud.service.ts [hunk 3]
```diff
@@ -379,4 +380,41 @@ export class TaskFilterCloudService extends BaseCloudService {
...
+     * If the new property is not found, it is created and assigned the value constructed from the old property.
+     * The filters are then updated in the preferences and returned.
+     * Old properties are left untouched for purposes like feature toggling.
+     *
+     * @param appName Name of the target app.
+     * @param key Key of the task filters.
+     * @param filters Array of task filters to be checked for backward compatibility.
+     * @returns Observable of task filters with updated properties.
+     */
+    private handleBackwardsCompatibility(appName: string, key: string, filters: TaskFilterCloudModel[]): Observable<TaskFilterCloudModel[]> {
```
Comment (BSekula) [inline lib/process-services-cloud/src/lib/task/task-filters/services/task-filter-cloud.service.ts:396]:
```suggestion
    private handleCreateFilterBackwardsCompatibility(appName: string, key: string, filters: TaskFilterCloudModel[]): Observable<TaskFilterCloudModel[]> {
```

File: lib/process-services-cloud/src/lib/task/task-list/services/task-list-cloud.service.ts [hunk 4]
```diff
@@ -56,6 +57,76 @@ export class TaskListCloudService extends BaseCloudService implements TaskListCl
...
+    }
+
+    getTaskListCounter(requestNode: TaskListRequestModel): Observable<number> {
+        if (!requestNode.appName) {
+            return throwError(() => new Error('Appname not configured'));
+        }
+        return this.fetchTaskList(requestNode).pipe(map((tasks) => tasks.list.pagination.totalItems));
+    }
+
+    protected buildQueryData(requestNode: TaskListRequestModel) {
```
Comment (BSekula) [inline lib/process-services-cloud/src/lib/task/task-list/services/task-list-cloud.service.ts:101]: can we infer return type here?

File: lib/process-services-cloud/src/lib/task/task-list/services/task-list-cloud.service.ts [hunk 5]
```diff
@@ -56,6 +57,76 @@ export class TaskListCloudService extends BaseCloudService implements TaskListCl
...
+            createdFrom: requestNode.createdFrom,
+            createdTo: requestNode.createdTo,
+            dueDateFrom: requestNode.dueDateFrom,
+            dueDateTo: requestNode.dueDateTo,
+            variableKeys
+        };
+
+        Object.keys(queryData).forEach((key) => {
+            const value = queryData[key];
+            if (value === undefined || value === null || value === '' || (Array.isArray(value) && (value.length === 0 || value[0] === null))) {
```
Comment (BSekula) [inline lib/process-services-cloud/src/lib/task/task-list/services/task-list-cloud.service.ts:122]:
```suggestion
           const isValueEmpty = !value;
           const isValueEmptyArray = Array.isArray(value) && (value.length === 0 || value[0] === null);

           if (isValueEmpty || isValueEmptyArray) {
```

**Review by BSekula — APPROVED** [review]

File: lib/process-services-cloud/src/lib/task/task-list/services/task-list-cloud.service.ts (hunk 4 above)
Comment (DudaRobert) [reply lib/process-services-cloud/src/lib/task/task-list/services/task-list-cloud.service.ts:101]: It's just an object - even in the `BaseCloudService.post` method it's `any`

**Review by eromano — APPROVED** [review]

**Pull request merged**